EIPS/eip-5058.md

```diff
@@ -0,0 +1,205 @@
+    /**
+     * @dev Emitted when `owner` enables `approved` to lock the `tokenId` token.
+     */
+    event LockApproval(address indexed owner, address indexed approved, uint256 indexed tokenId);
+
+    /**
+     * @dev Emitted when `owner` enables or disables (`approved`) `operator` to lock all of its tokens.
+     */
+    event LockApprovalForAll(address indexed owner, address indexed operator, bool approved);
+
+    /**
+     * @dev Returns the locker who is locking the `tokenId` token.
+     *
+     * Requirements:
+     *
+     * - `tokenId` must exist.
+     */
+    function lockerOf(uint256 tokenId) external view returns (address locker);
+
+    /**
+     * @dev Lock `tokenId` token until the block number is greater than `expired` to be unlocked.
+     *
+     * Requirements:
+     *
+     * - `from` cannot be the zero address.
+     * - `tokenId` token must be owned by `from`.
+     * - `expired` must be greater than block.number
+     * - If the caller is not `from`, it must be approved to lock this token
+     * by either {lockApprove} or {setLockApprovalForAll}.
+     *
+     * Emits a {Locked} event.
+     */
+    function lockFrom(
+        address from,
+        uint256 tokenId,
+        uint256 expired
+    ) external;
+
+    /**
+     * @dev Unlock `tokenId` token.
+     *
+     * Requirements:
+     *
+     * - `from` cannot be the zero address.
+     * - `tokenId` token must be owned by `from`.
+     * - the caller must be the operator who locks the token by {lockFrom}
+     *
+     * Emits a {Unlocked} event.
+     */
+    function unlockFrom(address from, uint256 tokenId) external;
+
+    /**
+     * @dev Gives permission to `to` to lock `tokenId` token.
+     *
+     * Requirements:
+     *
+     * - The caller must own the token or be an approved lock operator.
+     * - `tokenId` must exist.
+     *
+     * Emits an {LockApproval} event.
+     */
+    function lockApprove(address to, uint256 tokenId) external;
+
+    /**
+     * @dev Approve or remove `operator` as an lock operator for the caller.
+     * Operators can call {lockFrom} for any token owned by the caller.
+     *
+     * Requirements:
+     *
+     * - The `operator` cannot be the caller.
+     *
+     * Emits an {LockApprovalForAll} event.
+     */
+    function setLockApprovalForAll(address operator, bool approved) external;
+
+    /**
+     * @dev Returns the account lock approved for `tokenId` token.
+     *
+     * Requirements:
+     *
+     * - `tokenId` must exist.
+     */
+    function getLockApproved(uint256 tokenId) external view returns (address operator);
+
+    /**
+     * @dev Returns if the `operator` is allowed to lock all of the assets of `owner`.
+     *
+     * See {setLockApprovalForAll}
+     */
+    function isLockApprovedForAll(address owner, address operator) external view returns (bool);
+
+    /**
+     * @dev Returns if the `tokenId` token is locked.
+     */
+    function isLocked(uint256 tokenId) external view returns (bool);
```

Comment:
Also don't we need to get the expiration block number ?

Reply:
You are right, I will add the expiration block time parameter.

Reply:
add `lockedExpiredTime` function

Reply:
Thanks for adding this. Please resolve this comment. I don't have a permission to do that.